# IFQ718 Assignment 2

---

## Questions

Select any three

1. What is the profit of each business per year?

2. Which product generated the most profit for each business per year?

3. What was the most popular product for each business in each year?

4. Are there periods of the year where some businesses are more profitable?

5. What were the most utilised promotions for each business?

6. Which customers were most loyal for each business?

7. What is the total expenditure of loyal customers compared to one-off customers?

8. What is the employee turnover rate of each business?

9. How much has inflation impacted the profit margin of each business?

10. What is the impact of seasonal fluctuations on each business?

---

## Getting started

We have provided some code below to get you started

In [3]:
# Run this cell to begin working with this Notebook
# Please make any modifications to suit your needs

import os
import csv
import json
import glob
import json
from collections import defaultdict
from datetime import datetime

---

# Preparations

If you wish, you may prepare frames or files prior to the Exploration sections.

For example, if you want to prepare a frame that is used in multiple Exploration sectons, you may prepare it here.

In [4]:
# Function to load receipts from JSON files
def load_receipts(directory):
    receipts = []
    for file_path in glob.glob(f'{directory}/*.json'):
        with open(file_path, 'r') as file:
            receipt = json.load(file)
            receipts.append(receipt)
    return receipts

# Main script
directory = 'receipts-json'
receipts = load_receipts(directory)

---

## Exploration One

### Introduction

Be sure to make your aim/question clear.

Understanding the annual profit of a business is essential for evaluating its financial health and sustainability. Our aim is to calculate the net profit of each business per year. By analysing these annual profits, businesses can make informed decisions on investments and strategies for growth. This analysis helps in assessing current performance and planning for future profitability. However, it's important to note that initial data often reflects only the net profit, not the gross profit. To gain a more comprehensive understanding of the financial situation, the next step should be to subtract all costs from the revenue to calculate the gross profit. This approach provides a clearer picture of the business's financial dynamics and operational efficiency.

### Methods

In [5]:
# Function to calculate revenue per business per year
def calculate_revenue(receipts):
    revenue_per_year = defaultdict(lambda: defaultdict(float))
    for receipt in receipts:
        business_name = receipt["Business"]["Name"]
        date_str = receipt["Date"]
        total_amount = receipt["Total"]
        date_obj = datetime.strptime(date_str, "%Y/%m/%d")
        year = date_obj.year
        revenue_per_year[business_name][year] += total_amount
    return revenue_per_year

# Calculate revenue
revenue_per_year = calculate_revenue(receipts)
print("Revenue per Business per Year:")
for business, yearly_revenue in revenue_per_year.items():
    print(f"Business: {business}")
    for year in sorted(yearly_revenue):
        total_revenue = yearly_revenue[year]
        print(f"  Year: {year} - Gross Net Profit: ${total_revenue:.2f}")

Revenue per Business per Year:
Business: Ed's Barber Supplies
  Year: 2022 - Gross Net Profit: $15843.00
  Year: 2023 - Gross Net Profit: $35541.62
  Year: 2024 - Gross Net Profit: $47132.80
  Year: 2025 - Gross Net Profit: $55789.68
  Year: 2026 - Gross Net Profit: $69863.00
  Year: 2027 - Gross Net Profit: $59039.85
  Year: 2028 - Gross Net Profit: $58482.48
  Year: 2029 - Gross Net Profit: $29114.76
Business: Wake Up with Coffee
  Year: 2022 - Gross Net Profit: $8752.00
  Year: 2023 - Gross Net Profit: $14507.15
  Year: 2024 - Gross Net Profit: $21825.22
  Year: 2025 - Gross Net Profit: $20013.75
  Year: 2026 - Gross Net Profit: $16829.84
  Year: 2027 - Gross Net Profit: $25871.98
  Year: 2028 - Gross Net Profit: $26398.32
  Year: 2029 - Gross Net Profit: $11558.98
Business: Penguin Swim School
  Year: 2022 - Gross Net Profit: $6650.00
  Year: 2023 - Gross Net Profit: $24567.98
  Year: 2024 - Gross Net Profit: $24051.95
  Year: 2025 - Gross Net Profit: $29405.12
  Year: 2026 - Gross

### Discussion

What did you find out? You may make reference to or include small portions of the frames generated in your methods.

---

## Exploration Two

### Introduction

What was the most popular product for each business in each year?

Understanding the popularity of products is crucial for evaluating a business's market performance and consumer preferences. Our goal is to identify the most popular product for each business each year. By analysing these preferences, businesses can make informed decisions on inventory management, marketing strategies, and product development. This analysis aids in assessing current consumer trends and planning for future product offerings. However, initial data often reflects only sales volume, not the profitability of each product. 

### Methods

In [6]:
# Function to find the most popular product per business per year
def calculate_product_sales(receipts):
    product_sales = defaultdict(lambda: defaultdict(lambda: defaultdict(int)))
    for receipt in receipts:
        business_name = receipt["Business"]["Name"]
        date_str = receipt["Date"]
        date_obj = datetime.strptime(date_str, "%Y/%m/%d")
        year = date_obj.year
        for product in receipt.get("Products", []):
            product_name = product["Name"]
            quantity_sold = product["Quantity"]
            product_sales[business_name][year][product_name] += quantity_sold
    return product_sales

# Function to find the most popular product for each business per year
def find_most_popular_products(product_sales):
    most_popular_products = {}
    for business, yearly_sales in product_sales.items():
        most_popular_products[business] = {}
        for year, products in yearly_sales.items():
            most_popular_product = max(products, key=products.get)
            most_popular_products[business][year] = (most_popular_product, products[most_popular_product])
    return most_popular_products


# Calculate product sales and find most popular products
product_sales = calculate_product_sales(receipts)
most_popular_products = find_most_popular_products(product_sales)
print("\nMost Popular Products per Business per Year:")
for business, yearly_products in most_popular_products.items():
    print(f"Business: {business}")
    for year in sorted(yearly_products):
        product, quantity = yearly_products[year]
        print(f"  Year: {year} - Most Popular Product: {product} (Quantity Sold: {quantity})")


Most Popular Products per Business per Year:
Business: Ed's Barber Supplies
  Year: 2022 - Most Popular Product: Soap 500mL (Quantity Sold: 171)
  Year: 2023 - Most Popular Product: Soap 500mL (Quantity Sold: 334)
  Year: 2024 - Most Popular Product: Razorblade (Quantity Sold: 309)
  Year: 2025 - Most Popular Product: Broom (Quantity Sold: 358)
  Year: 2026 - Most Popular Product: Cape white (Quantity Sold: 384)
  Year: 2027 - Most Popular Product: Soap 500mL (Quantity Sold: 372)
  Year: 2028 - Most Popular Product: Shaving cream (Quantity Sold: 382)
  Year: 2029 - Most Popular Product: Spray bottle (Quantity Sold: 221)
Business: Wake Up with Coffee
  Year: 2022 - Most Popular Product: Mocha (Quantity Sold: 200)
  Year: 2023 - Most Popular Product: Ham toastie (Quantity Sold: 402)
  Year: 2024 - Most Popular Product: Ham bagel (Quantity Sold: 440)
  Year: 2025 - Most Popular Product: Salmon bagel (Quantity Sold: 481)
  Year: 2026 - Most Popular Product: Long Black (Quantity Sold: 542)

### Discussion

What did you find out? You may make reference to or include small portions of the frames generated in your methods.

The significant decline in sales at Penguin Swim School in 2029, especially evident with Aquaerobics sales dropping to just 7 units from higher figures in previous years, can be attributed to various factors. Meanwhile, Ed's Barber Supplies consistently sees Soap 500mL as its top-selling product across multiple years, with Rasorblade and Cape white also enjoying periods of high popularity. At Wake Up with Coffee, Mocha led sales in 2022, followed by shifts to Ham toastie, Ham bagel, Long Black, and Cupcake in subsequent years, reflecting changing consumer preferences and product trends. 

---

## Exploration Three

### Introduction

Be sure to make your aim/question clear

Which customers were most loyal for each business?

Understanding customer loyalty is essential for evaluating a business's customer retention and satisfaction. Our objective is to identify the most loyal customers for each business. By examining loyalty patterns, businesses can make informed decisions on managing customer relationships, developing marketing strategies, and improving services. This analysis helps recognise repeat business trends and plan effective customer retention initiatives. However, initial data often focuses solely on purchase frequency, overlooking the overall value each loyal customer contributes. To gain a more comprehensive view of customer loyalty, it's crucial to consider both the frequency and monetary value of their purchases. This method provides deeper insights into the business's customer dynamics and operational efficiency.

### Methods

In [7]:
# Function to find the most loyal customer per business
def calculate_customer_loyalty(receipts):
    visit_count = defaultdict(lambda: defaultdict(int))
    for receipt in receipts:
        business_name = receipt["Business"]["Name"]
        customer_name = receipt.get("Customer", {}).get("Name", "Unknown Customer")
        if customer_name != "Unknown Customer":
            visit_count[business_name][customer_name] += 1
    return visit_count


# Function to find the most loyal customers for each business
def find_most_loyal_customers(visit_count):
    most_loyal_customers = {}
    for business, customers in visit_count.items():
        max_visits = max(customers.values())
        most_loyal_customers[business] = [customer for customer, count in customers.items() if count == max_visits]
    return most_loyal_customers


# Calculate customer loyalty and find most loyal customers
visit_count = calculate_customer_loyalty(receipts)
most_loyal_customers = find_most_loyal_customers(visit_count)
print("\nMost Loyal Customers per Business:")
for business, loyal_customers in most_loyal_customers.items():
    print(f"Business: {business}")
    for customer in loyal_customers:
        print(f"  Most Loyal Customer: {customer} - Visits: {visit_count[business][customer]}")


Most Loyal Customers per Business:
Business: Wake Up with Coffee
  Most Loyal Customer: Scott - Visits: 181
Business: Ed's Barber Supplies
  Most Loyal Customer: Lisa - Visits: 99
Business: Penguin Swim School
  Most Loyal Customer: Scott - Visits: 83


### Discussion

What did you find out? You may make reference to or include small portions of the frames generated in your methods.


The loyalty data provides intriguing insights into customer behavior at Wake Up with Coffee, Ed's Barber Supplies, and Penguin Swim School. Scott emerges as the most loyal customer for both Wake Up with Coffee (181 visits) and Penguin Swim School (83 visits), with visit counts that significantly exceed those of other customers. This discrepancy suggests potential data entry anomalies rather than reflecting typical customer behavior. Lisa, with 99 visits, stands out as the top patron for Ed's Barber Supplies, demonstrating considerable loyalty.  Examining customer visit counts reveals a mix of outliers and consistent visitation patterns. While Scott's exceptionally high visit frequencies may raise questions about data accuracy, many other customers maintain regular visit counts ranging from the 80s to over 100 at Wake Up with Coffee and Ed's Barber Supplies, and in the 40s and 50s at Penguin Swim School. This indicates a core group of regular customers and robust loyalty across these businesses. Overall, the customer loyalty at these businesses is strong. These findings underscore the importance of accurately recording and analysing customer data to inform effective business strategies. Addressing potential anomalies in loyalty data, such as Scott's unusually high visit counts, can lead to better understanding of customer behavior and more precise targeting of loyalty programs. By leveraging these insights, businesses can optimise marketing efforts, enhance customer satisfaction, and foster long-term customer relationships.